haskell/experimental/defs.bzl

```diff
@@ -0,0 +1,127 @@
+        ),
+        # TODO[AH] Support package name and version for modules that are part of a package.
+        # TODO[AH] Suppport worker
+    },
+    toolchains = [
+        "@bazel_tools//tools/cpp:toolchain_type",
+        "@rules_haskell//haskell:toolchain",
+        "@rules_sh//sh/posix:toolchain_type",
+    ],
+    fragments = ["cpp"],
+)
+
+def haskell_module(
+        name,
+        src = None,
+        extra_srcs = [],
+        deps = [],
+        ghcopts = [],
+        #repl_ghci_args = [],
+        plugins = [],
+        tools = [],
+        worker = None,
+        **kwargs):
+    """Compile a module from Haskell source.
+
+    Note: This rule is experimental and not ready for production, yet.
+
+    ### Examples
+
+      ```bzl
+      haskell_module(
+          name = "Example.Module",
+          src = "src/Example/Module.hs",
+          src_strip_prefix = "src",
+          deps = [
+              "//:Another.Module",
+              "//:some-library",
+          ],
+      )
+      ```
+
+    Args:
+      name: A unique name for this rule.
+      src_strip_prefix: Prefix before the path matches the module name.
+        This is used as an import search for the Haskell compiler.
+        Values starting with `/` are relative to the workspace root,
+        other paths are relative to the package.
+      src: The Haskell source file.
+      extra_srcs: Extra (non-Haskell) source files that will be needed at compile time (e.g. by Template Haskell).
+      deps: List of other Haskell modules or libraries needed to compile this module.
+      data: See [Bazel documentation](https://docs.bazel.build/versions/master/be/common-definitions.html#common.data).,
+      ghcopts: Flags to pass to Haskell compiler. Subject to Make variable substitution.
+      plugins: Compiler plugins to use during compilation. (Not implemented, yet)
+      tools: Extra tools needed at compile-time, like preprocessors. (Not implemented, yet)
+      worker: Experimental. Worker binary employed by Bazel's persistent worker mode. See [use-cases documentation](https://rules-haskell.readthedocs.io/en/latest/haskell-use-cases.html#persistent-worker-mode-experimental). (Not implemented, yet)
+      **kwargs: Common rule attributes. See [Bazel documentation](https://docs.bazel.build/versions/master/be/common-definitions.html#common-attributes).
+    """
+    _haskell_module(
+        name = name,
+        src = src,
+        extra_srcs = extra_srcs,
+        deps = deps,
+        ghcopts = ghcopts,
+        #repl_ghci_args = repl_ghci_args,
+        plugins = plugins,
+        tools = tools,
+        #worker = worker,
+        **kwargs
+    )
+
+    #repl_kwargs = {
+    #    attr: kwargs[attr]
+    #    for attr in ["testonly", "tags"]
+    #    if attr in kwargs
+    #}
+    #haskell_repl(
+    #    name = "%s@repl" % name,
+    #    deps = [name],
+    #    experimental_from_source = [":%s" % name],
+    #    repl_ghci_args = [],
+    #    **repl_kwargs
+    #)
```

Comment:
Why are these here but commented? Presumably they don't work yet but should be added later?

Reply:
Yes, I followed the pattern of `haskell_library` and left this part commented since it doesn't work, yet. However, there's no good reason to keep it around. I've removed it.

Reply:
Oh, I didn't realize this part of the code was also taken from there. I should have checked.